### ISSUE TITLE
Observers are lost after first Unit test runs.

### ISSUE THREAD

**Issue by ClickerMonkey** [laravel/framework#7882]
I have a trait that looks like this:

```php
/**
 * A model which stores which user created it.
 */
trait HasCreatedBy
{

  /**
   * Returns the user who created this model.
   */
  public function createdBy()
  {
    return $this->belongsTo('\App\User', $this->createdByColumn());
  }
  
  /**
   * The column which is a key to the user who created this model.
   *
   * This can be overriden by any class that includes this trait.
   */
  public function createdByColumn()
  {
    return 'created_by';
  }
  
  /**
   * Sets the created_by property before the model is inserted.
   */
  public static function bootHasCreatedBy()
  {
    static::creating(function($model) 
    {
      $column = $model->createdByColumn();
      
      $model->setCreatedAt( $model->freshTimestamp() );
      $model->setAttribute( $column, \Auth::id() );
      
    }, Priority::HasCreatedBy);
  }
}
```
For every model I have that has created_by & created_at I include this trait and it handles setting the fields before the model is inserted (I have many traits just like this one).

The issue is that between each Unit Test the application is rebuilt which clears the listeners/Dispatcher. The second test that runs hits none of the listeners that were previously loaded - which is a problem.

I tried running `phpunit --process-isolation` to remedy this problem but that doesn't appear to work with Laravel, I get this error:

```
PHPUnit_Framework_Exception: PHP Notice:  Constant LARAVEL_START already defined in /Users/phil/Sites/timetally2/bootstrap/autoload.php on line 3
```

TLDR; After first test all registered events are lost but since PHP classes are already loaded the events aren't reregistered between tests.

Suggestions?

**Comment by ClickerMonkey** [comment]
My current solution, although I'm sure it might cause problems somewhere else, is to only have a single instance of the application and avoid flushing it. I modified the createApplication in the `tests/TestCase` file:

```php
	// the single application instance
	private static $_app = null;

	/**
	 * Creates the application.
	 *
	 * @return \Illuminate\Foundation\Application
	 */
	public function createApplication()
	{    
    	if (self::$_app === null)
    	{      
  			self::$_app = require __DIR__.'/../bootstrap/app.php';
  			self::$_app->make('Illuminate\Contracts\Console\Kernel')->bootstrap();
    	}
      
		return self::$_app;
	}
```

And I had to modify `Illuminate\Foundation\Testing\TestCase` to comment out the flush:

```php
	/**
	 * Clean up the testing environment before the next test.
	 *
	 * @return void
	 */
	public function tearDown()
	{
		if ($this->app)
		{
			//$this->app->flush();
		}
	}
```

This is a fine temporary solution, but I'm sure this will cause issues later.

**Comment by crynobone** [comment]
It has been discussed before:

https://github.com/laravel/framework/issues/6785#issuecomment-68017524
https://github.com/laravel/framework/issues/1181#issuecomment-18099041

**Comment by ClickerMonkey** [comment]
Ah thanks, I need to dig a little deeper next time.

Unfortunately none of the related issues address the problem, they all say you need to register events a different way which makes my code way less flexible and more coupled... I don't like thay!
